jobs/integration/test_opa.py

```diff
@@ -0,0 +1,439 @@
+        while total_violations == 0:
+            action = await unit.run_action("list-violations")
+            await action.wait()
+            assert action.status == "completed"
+            assert "constraint-violations" in action.results
+            constraint_violations = json.loads(action.results["constraint-violations"])
+            if constraint_violations:
+                total_violations = constraint_violations[0].get("total-violations")
+            await asyncio.sleep(5)
+
+        log("Running get-violation action")
+        action = await unit.run_action(
+            "get-violation",
+            **{
+                "constraint-template": "K8sRequiredLabels",
+                "constraint": "ns-must-have-gk",
+            },
+        )
+        await action.wait()
+        assert len(json.loads(action.results["violations"])) > 0
+
+    async def wait_for_units(
+        self, app, error_status=("blocked", "error"), expected_status="active"
+    ):
+        while (
+            status := app.units[0].workload_status if app.units else None
+        ) != expected_status:
+            assert status not in error_status
+            await asyncio.sleep(5)
+
+    async def wait_for_app_removed(self, app_name, k8s_model=None):
+        if not k8s_model:
+            k8s_model = self.k8s_model
+
+        async def check_app_missing():
+            apps = await k8s_model.get_status()
+            return app_name not in apps.applications
+
+        try:
+            await block_until_with_coroutine(check_app_missing, timeout=120)
+        except asyncio.TimeoutError:
+            raise AssertionError(f"Application {app_name} was not removed")
+
+
+@pytest.mark.clouds(["vsphere"])
+class TestOPAWebhook(OPATestBase):
+    async def test_opa_webhook_ready(self, opa_controller_manager):
+        await self.wait_for_units(opa_controller_manager)
+
+        log("Waiting for gatekeeper charm to be ready")
+        await wait_for_application_status(
+            self.k8s_model, "gatekeeper-controller-manager", status="active"
+        )
+
+    async def test_create_ns_no_policy(self, opa_controller_manager):
+        try:
+            await self.validate_create_ns()
+        except JujuRunError as e:
+            if e.output.startswith(
+                "Error from server (InternalError): Internal error occurred"
+            ):
+                # Try again as it might be transient/initialization related
+                await asyncio.sleep(5)
+                await self.validate_create_ns()
+            else:
+                raise
+
+    async def test_policy_enforcement(self, opa_controller_manager):
+        log("Creating policy and constraint crds")
+        await self.deploy_example_policy()
+        log("Test that the policy is enforced")
+        try:
+            await self.validate_create_ns_fail()
+            await self.validate_create_ns_with_label()
+        finally:
+            await self.destroy_example_policy()
```

Comment:
this test doesn't assert anything?  Are we relying on the fact that nothing is raised in these `validate` methods?  As long as they have asserts i guess that's okay

Reply:
the `validate_*` assert that the correct errors are returned (or not returned)